app/src/debugger.lisp

```diff
@@ -0,0 +1,299 @@
+
+(defstruct command
+  (name nil :type string)
+  (function nil :type function)
+  (documentation nil :type string))
+
+(defparameter *commands* nil
+  "Association list of available commands.")
+
+(defmacro define-debugger-command (name lambda-list &body body)
+  "Define a new command function named %NAME (note that the % character is prepended to NAME) with the specified LAMBDA-LIST and BODY."
+  (unless (stringp (first body))
+    (error "Documentation string missing."))
+  (let ((documentation (pop body))
+        (lowercase-name (string-downcase name))
+        (symbol (intern (concatenate 'string "%" (symbol-name name)))))
+    `(progn
+       (defun ,symbol ,lambda-list
+         ,@body)
+
+       (setf *commands* (acons ,lowercase-name
+                               (make-command :name ,lowercase-name
+                                             :function (symbol-function ',symbol)
+                                             :documentation ,documentation)
+                               *commands*)))))
+
+(defun string-prefix-p (a b)
+  (string= a b :end2 (min (length a) (length b))))
+
+(defun %get-command (command-name)
+  (let ((match (assoc command-name *commands* :test #'string-prefix-p)))
+    (when match
+      (rest match))))
+
+(defun get-command (command-name)
+  "Return COMMAND structure corresponding to COMMAND-NAME. Signal an error if no such command exists."
+  (let ((command (%get-command command-name)))
+    (unless command
+      (error "No such command ~S exists." command-name))
+    command))
+
+(defun command-exists-p (command-name)
+  (not (null (%get-command command-name))))
+
+(defun eval-command (command-name &optional args)
+  (let ((function (command-function (get-command command-name))))
+    (if (string= args "")
+        (funcall function)
+        (funcall function args))))
+
+(defun command-description (command)
+  (aref (nth-value 1 (cl-ppcre:scan-to-strings ".*\\n\\n(.*)" (command-documentation command))) 0))
+
+(define-debugger-command help (&optional command-name)
+  "Usage: help [COMMAND]
+
+Display help message."
+  (if command-name
+      (format t "~A~%" (command-documentation (get-command command-name)))
+      (loop :for (command-name . command) :in *commands* :do
+        (format t "~&~A~40T~A~%" command-name (command-description command)))))
+
+(define-debugger-command reset ()
+  "Usage: reset
+
+Reset the current QVM to the zero ket and set the program counter to zero."
+  (when *qvm*
+    (qvm::reset-quantum-state *qvm*)
+    (qvm::reset-classical-memory *qvm*)
+    (setf (qvm::pc *qvm*) 0)))
+
+(define-debugger-command list ()
+  "Usage: list
+
+Disassemble (list) the currently loaded program."
+  (unless *qvm*
+    (error "No program has been loaded."))
+  (loop :for instruction-number :from 0
+        :for instruction :across (qvm::program *qvm*) :do
+          (format t "~&~D~10T~/cl-quil:instruction-fmt/~%"
+                  instruction-number instruction)))
```

Comment:
would be cool to indicate the current location somehow, e.g. when `(= (qvm::pc *qvm*) instruction-number)`